serving/src/main/java/feast/serving/service/BigTableFeatureStorage.java

```diff
@@ -168,84 +123,35 @@ public BigTableFeatureStorage(Connection connection) {
-
-    List<FeatureValue> features = new ArrayList<>();
-    Map<String, Integer> featureCount = new HashMap<>(); // map of entity ID to number of feature.
-    try (Table table = connection.getTable(TableName.valueOf(entityName));
-        ResultScanner scanResult = table.getScanner(scan)) {
-      for (Result row : scanResult) {
-        Cell currentCell = row.getColumnLatestCell(getColumnFamily(featureSpec), featureIdBytes);
-        if (currentCell == null) {
-          continue;
-        }
-
-        BigTableRowKey rowKey = BigTableRowKey.parseFrom(currentCell.getRowArray());
-        String entityId = rowKey.getEntityKey();
-        byte[] cellValue = currentCell.getValueArray();
-
-        if (cellValue == null) {
-          continue;
-        }
-
-        Integer count = featureCount.computeIfAbsent(entityId, k -> 0);
-        if (count < limit) {
-          Timestamp timestamp = Timestamps.fromMillis(currentCell.getTimestamp());
-          Value value = Value.parseFrom(cellValue);
-          FeatureValue featureValue = new FeatureValue(featureId, entityId, value, timestamp);
-          features.add(featureValue);
-
-          count++;
-          featureCount.put(entityId, count);
-        }
-      }
-
-      return features;
-    } catch (IOException e) {
-      log.error("Error while retrieving feature: {}", e);
-      throw new FeatureRetrievalException("Unable to retrieve feature from BigTable", e);
-    }
-  }
-
-  /**
-   * Create list of BigTable's Get operation.
+   * Create list of get operation for retrieving a feature of several entities optionally filtered
+   * by its timestamp.
    *
    * @param entityIds list of entity ID.
-   * @param featureIdBytes byte array value of a feature ID.
-   * @param columnFamily byte array value of column family
+   * @param featureSpec feature spec
+   * @param tsRange time filter
    * @return list of Get operation.
    */
   private List<Get> createGets(
-      Collection<String> entityIds, byte[] featureIdBytes, byte[] columnFamily) {
+      Collection<String> entityIds, FeatureSpec featureSpec, TimestampRange tsRange) {
+    byte[] featureIdBytes = featureSpec.getId().getBytes();
+    byte[] columnFamily = getColumnFamily(featureSpec);
+    long start = 0;
+    long end = 0;
+    if (TimeUtil.isTimeFilterRequired(tsRange)) {
+      start = Timestamps.toMillis(tsRange.getStart());
+      end = Timestamps.toMillis(tsRange.getEnd());
+    }
+
     List<Get> gets = new ArrayList<>();
     for (String entityId : entityIds) {
       String entityIdPrefix = DigestUtils.sha1Hex(entityId.getBytes()).substring(0, 7);
       BigTableRowKey btKey = createRowKey(entityIdPrefix, entityId, "0");
       Get get = new Get(btKey.toByteArray());
       get.addColumn(columnFamily, featureIdBytes);
       try {
-        // for some reason Get.readVersions has checked exception.
+        if (!featureSpec.getGranularity().equals(Enum.NONE) && tsRange != null) {
```

Comment:
We should not need this.

Reply:
Similarly to Redis, the we now only write to bigtable with a reversedMillis in the row key of "0", effectively always granularity of None, so there is no need to consider granularity at all.